README
NAME
    Encode::EUCJPMS - Microsoft Compatible Encodings for Japanese

SYNOPSIS
      use Encode::EUCJPMS;
      use Encode qw/encode decode/;
      $eucJP_ms = encode("eucJP-ms", $utf8);
      $utf8   = decode("eucJP-ms", $euc_jp);

ABSTRACT
    This module implements Microsoft compatible encodings for Japanese.
    Encodings supported are as follows.

      Canonical     Alias                                      Description
      --------------------------------------------------------------------
      eucJP-ms      qr/\beuc-?jp-?ms$/i                           eucJP-ms
                    qr/\beuc-?jp-?win$/i
      cp51932       qr/\bcp51932$/i                 Windows Codepage 51932
      cp50220                                       Windows Codepage 50220
      cp50221                                       Windows Codepage 50221
      --------------------------------------------------------------------

DESCRIPTION
    To find out how to use this module in detail, see Encode.

INSTALLATION
    To install this module type the following:

       perl Makefile.PL
       make
       make test
       make install

    If you want to add eucJP-ms to Encode's demand-loading list (so you
    don't have to "use Encode::EUCJPMS"), run

      enc2xs -C

    to update Encode::ConfigLocal, a module that controls local settings.
    After that, "use Encode;" is enough to load eucJP-ms on demand.

DEPENDENCIES
    This module requires perl version 5.7.3 or later.

AUTHOR
    NARUSE, Yui <naruse@airemix.com>

COPYRIGHT
    Copyright (C) 2005-2006 NARUSE, Yui <naruse@airemix.com>

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://www.perl.com/perl/misc/Artistic.html>

SEE ALSO
    Encode, Encode::JP

    Problems and Solutions for Unicode and User/Vendor Defined Characters
    <http://www.opengroup.or.jp/jvc/cde/ucs-conv-e.html>

    Windows Codepage 932
    <http://www.microsoft.com/globaldev/reference/dbcs/932.mspx>



Changes
#
# $Id: Changes 5 2006-06-23 17:26:15Z naruse $
# Revision history for Perl extension Encode::EUCJPMS.
#
$Revision: 5 $ $Date: 2006-06-24 02:26:15 +0900 (土, 24 6 2006) $
0.06  Sat Jun 24 02:00:00 2006
! EUCJPMS.pm
   Support CP20220 and CP20221
! ucm/eucJP-ms.ucm
   Import glibc-EUC_JP_MS-2.3.3.ucm
! ucm/cp51932.ucm
   Remove some illegal mappings.
! META.yaml
   Added.
0.05  Thu Oct 13 04:38:29 2005
! EUCJPMS.pm
   Fixed: qr/\bcp51932$/i was aliased to '"eucJP-ms"'
0.04  Tue Oct 11 03:59:38 2005
! ucm/eucJP-ms.ucm
  Fixed: Can't convert User defined characters, multi defined characters,
   JIS X 0212 without IBM extended characters in eucJP-ms to Unicode
   http://msyk.at.webry.info/200509/article_5.html
! ucm/cp51932.ucm
   Don't convert PUA (U+E000～U+E757) in Unicode to cp51932 convertion
   http://msyk.at.webry.info/200509/article_5.html
0.03  Fri Sep 09 10:56:50 2005
! ucm/cp51932.ucm
   Added.
0.02  Thu Sep 01 03:19:11 2005
! various mistakes are fixed
0.01  Thu Sep 01 03:19:11 2005
! First Release


Makefile.PL
#
# This file is auto-generated by:
# enc2xs version 2.01
# Fri Sep  9 10:16:57 2005
#
use 5.7.2;
use strict;
use ExtUtils::MakeMaker;
use Config;

# Please edit the following to the taste!
my $name = 'EUCJPMS';
my %tables = (
	     EUCJPMS_t   => [ 'ucm/eucJP-ms.ucm','ucm/cp51932.ucm' ],
	     );

#### DO NOT EDIT BEYOND THIS POINT!
require File::Spec;
my ($enc2xs, $encode_h) = ();
PATHLOOP:
for my $d (@Config{qw/bin sitebin vendorbin/}, 
	   (split /$Config{path_sep}/o, $ENV{PATH})){
    for my $f (qw/enc2xs enc2xs5.7.3/){
        my $path = File::Spec->catfile($d, $f);
        -r $path and $enc2xs = $path and last PATHLOOP;
    }
}
$enc2xs or die "enc2xs not found!";
print "enc2xs is $enc2xs\n";
my %encode_h = ();
for my $d (@INC){
    my $dir = File::Spec->catfile($d, "Encode");
    my $file =  File::Spec->catfile($dir, "encode.h");
    -f $file and $encode_h{$dir} = -M $file;
}
%encode_h or die "encode.h not found!";
# find the latest one
($encode_h) = sort {$encode_h{$b} <=> $encode_h{$a}} keys %encode_h;
print "encode.h is at $encode_h\n";

WriteMakefile(
              INC		=> "-I$encode_h",
	      NAME		=> 'Encode::'.$name,
	      VERSION_FROM	=> "$name.pm",
	      OBJECT		=> '$(O_FILES)',
	      'dist'		=> {
		  COMPRESS	=> 'gzip -9f',
		  SUFFIX	=> 'gz',
		  DIST_DEFAULT => 'all tardist',
	      },
	      MAN3PODS	=> {},
	      PREREQ_PM => {
			    'Encode'     => "1.41",
                           },
	      # OS 390 winges about line numbers > 64K ???
	      XSOPT => '-nolinenumbers',
	      );

package MY;

sub post_initialize
{
    my ($self) = @_;
    my %o;
    my $x = $self->{'OBJ_EXT'};
    # Add the table O_FILES
    foreach my $e (keys %tables)
    {
	$o{$e.$x} = 1;
    }
    $o{"$name$x"} = 1;
    $self->{'O_FILES'} = [sort keys %o];
    my @files = ("$name.xs");
    $self->{'C'} = ["$name.c"];
    # The next two lines to make MacPerl Happy -- dankogai via pudge
    $self->{SOURCE} .= " $name.c"
        if $^O eq 'MacOS' && $self->{SOURCE} !~ /\b$name\.c\b/;
    # $self->{'H'} = [$self->catfile($self->updir,'encode.h')];
    my %xs;
    foreach my $table (keys %tables) {
	push (@{$self->{'C'}},"$table.c");
	# Do NOT add $table.h etc. to H_FILES unless we own up as to how they
	# get built.
	foreach my $ext (qw($(OBJ_EXT) .c .h .exh .fnm)) {
	    push (@files,$table.$ext);
	}
    }
    $self->{'XS'} = { "$name.xs" => "$name.c" };
    $self->{'clean'}{'FILES'} .= join(' ',@files);
    open(XS,">$name.xs") || die "Cannot open $name.xs:$!";
    print XS <<'END';
#include <EXTERN.h>
#include <perl.h>
#include <XSUB.h>
#define U8 U8
#include "encode.h"
END
    foreach my $table (keys %tables) {
	print XS qq[#include "${table}.h"\n];
    }
    print XS <<"END";

static void
Encode_XSEncoding(pTHX_ encode_t *enc)
{
 dSP;
 HV *stash = gv_stashpv("Encode::XS", TRUE);
 SV *sv    = sv_bless(newRV_noinc(newSViv(PTR2IV(enc))),stash);
 int i = 0;
 PUSHMARK(sp);
 XPUSHs(sv);
 while (enc->name[i])
  {
   const char *name = enc->name[i++];
   XPUSHs(sv_2mortal(newSVpvn(name,strlen(name))));
  }
 PUTBACK;
 call_pv("Encode::define_encoding",G_DISCARD);
 SvREFCNT_dec(sv);
}

MODULE = Encode::$name	PACKAGE = Encode::$name
PROTOTYPES: DISABLE
BOOT:
{
END
    foreach my $table (keys %tables) {
	print XS qq[#include "${table}.exh"\n];
    }
    print XS "}\n";
    close(XS);
    return "# Built $name.xs\n\n";
}

sub postamble
{
    my $self = shift;
    my $dir  = "."; # $self->catdir('Encode');
    my $str  = "# $name\$(OBJ_EXT) depends on .h and .exh files not .c files - but all written by enc2xs\n";
    $str    .= "$name.c : $name.xs ";
    foreach my $table (keys %tables)
    {
	$str .= " $table.c";
    }
    $str .= "\n\n";
    $str .= "$name\$(OBJ_EXT) : $name.c\n\n";

    foreach my $table (keys %tables)
    {
	my $numlines = 1;
	my $lengthsofar = length($str);
	my $continuator = '';
	$str .= "$table.c : Makefile.PL";
	foreach my $file (@{$tables{$table}})
	{
	    $str .= $continuator.' '.$self->catfile($dir,$file);
	    if ( length($str)-$lengthsofar > 128*$numlines )
	    {
		$continuator .= " \\\n\t";
		$numlines++;
	    } else {
		$continuator = '';
	    }
	}
	my $plib   = $self->{PERL_CORE} ? '"-I$(PERL_LIB)"' : '';
	my $ucopts = '-"Q"';
	$str .=  
	    qq{\n\t\$(PERL) $plib $enc2xs $ucopts -o \$\@ -f $table.fnm\n\n};
	open (FILELIST, ">$table.fnm")
	    || die "Could not open $table.fnm: $!";
	foreach my $file (@{$tables{$table}})
	{
	    print FILELIST $self->catfile($dir,$file) . "\n";
	}
	close(FILELIST);
    }
    return $str;
}



EUCJPMS.pm
#
# $Id: EUCJPMS.pm 5 2006-06-23 17:26:15Z naruse $
#
package Encode::EUCJPMS;
use strict;
our $VERSION = "0.06";
 
use Encode qw(:fallbacks);
use XSLoader;
XSLoader::load(__PACKAGE__,$VERSION);

Encode::define_alias(qr/\beuc-?jp-?ms$/i =>  '"eucJP-ms"');
Encode::define_alias(qr/\beuc-?jp-?win$/i => '"eucJP-ms"');

for my $name ('cp50220','cp50221'){ #, 'cp50222'
    my $h2z     = ($name eq 'cp50220')    ? 1 : 0;
    my $jis0212 = 0;

    $Encode::Encoding{$name} =
        bless {
               Name      =>   $name,
               h2z       =>   $h2z,
               jis0212   =>   $jis0212,
              } => __PACKAGE__;
}

use base qw(Encode::Encoding);

# we override this to 1 so PerlIO works
sub needs_lines { 1 }

use Encode::CJKConstants qw(:all);

#
# decode is identical for all 2022 variants
#

sub decode($$;$)
{
    my ($obj, $str, $chk) = @_;
    my $residue = '';
    if ($chk){
	$str =~ s/([^\x00-\x7f].*)$//so and $residue = $1;
    }
    $residue .= jis_euc(\$str);
    $_[1] = $residue if $chk;
    return Encode::decode('cp51932', $str, FB_PERLQQ);
}

#
# encode is different
#

sub encode($$;$)
{
    require Encode::JP::H2Z;
    my ($obj, $utf8, $chk) = @_;
    # empty the input string in the stack so perlio is ok
    $_[1] = '' if $chk;
    my ($h2z, $jis0212) = @$obj{qw(h2z jis0212)};
    my $octet = Encode::encode('cp51932', $utf8, FB_PERLQQ) ;
    $h2z and &Encode::JP::H2Z::h2z(\$octet);
    euc_jis(\$octet, $jis0212);
    return $octet;
}

#
# cat_decode
#
my $re_scan_jis_g = qr{
   \G ( ($RE{JIS_0212}) |  $RE{JIS_0208}  |
        ($RE{ISO_ASC})  | ($RE{JIS_KANA}) | )
      ([^\e]*)
}x;
sub cat_decode { # ($obj, $dst, $src, $pos, $trm, $chk)
    my ($obj, undef, undef, $pos, $trm) = @_; # currently ignores $chk
    my ($rdst, $rsrc, $rpos) = \@_[1,2,3];
    local ${^ENCODING};
    use bytes;
    my $opos = pos($$rsrc);
    pos($$rsrc) = $pos;
    while ($$rsrc =~ /$re_scan_jis_g/gc) {
	my ($esc, $esc_0212, $esc_asc, $esc_kana, $chunk) =
	  ($1, $2, $3, $4, $5);

	unless ($chunk) { $esc or last;  next; }

	if ($esc && !$esc_asc) {
	    $chunk =~ tr/\x21-\x7e/\xa1-\xfe/;
	    if ($esc_kana) {
		$chunk =~ s/([\xa1-\xdf])/\x8e$1/og;
	    } elsif ($esc_0212) {
		$chunk =~ s/([\xa1-\xfe][\xa1-\xfe])/\x8f$1/og;
	    }
	    $chunk = Encode::decode('cp51932', $chunk, 0);
	}
	elsif ((my $npos = index($chunk, $trm)) >= 0) {
	    $$rdst .= substr($chunk, 0, $npos + length($trm));
	    $$rpos += length($esc) + $npos + length($trm);
	    pos($$rsrc) = $opos;
	    return 1;
	}
	$$rdst .= $chunk;
	$$rpos = pos($$rsrc);
    }
    $$rpos = pos($$rsrc);
    pos($$rsrc) = $opos;
    return '';
}

# JIS<->EUC
my $re_scan_jis = qr{
   (?:($RE{JIS_0212})|$RE{JIS_0208}|($RE{ISO_ASC})|($RE{JIS_KANA}))([^\e]*)
}x;

sub jis_euc {
    local ${^ENCODING};
    my $r_str = shift;
    $$r_str =~ s($re_scan_jis)
    {
	my ($esc_0212, $esc_asc, $esc_kana, $chunk) =
	   ($1, $2, $3, $4);
	if (!$esc_asc) {
	    $chunk =~ tr/\x21-\x7e/\xa1-\xfe/;
	    if ($esc_kana) {
		$chunk =~ s/([\xa1-\xdf])/\x8e$1/og;
	    }
	    elsif ($esc_0212) {
		$chunk =~ s/([\xa1-\xfe][\xa1-\xfe])/\x8f$1/og;
	    }
	}
	$chunk;
    }geox;
    my ($residue) = ($$r_str =~ s/(\e.*)$//so);
    return $residue;
}

sub euc_jis{
    no warnings qw(uninitialized);
    my $r_str = shift;
    my $jis0212 = shift;
    $$r_str =~ s{
	((?:$RE{EUC_C})+|(?:$RE{EUC_KANA})+|(?:$RE{EUC_0212})+)
	}{
	    my $chunk = $1;
	    my $esc =
		( $chunk =~ tr/\x8E//d ) ? $ESC{KANA} :
		    ( $chunk =~ tr/\x8F//d ) ? $ESC{JIS_0212} :
			$ESC{JIS_0208};
	    if ($esc eq $ESC{JIS_0212} && !$jis0212){
		# fallback to '?'
		$chunk =~ tr/\xA1-\xFE/\x3F/;
	    }else{
		$chunk =~ tr/\xA1-\xFE/\x21-\x7E/;
	    }
	    $esc . $chunk . $ESC{ASC};
	}geox;
    $$r_str =~
	s/\Q$ESC{ASC}\E
	    (\Q$ESC{KANA}\E|\Q$ESC{JIS_0212}\E|\Q$ESC{JIS_0208}\E)/$1/gox;
    $$r_str;
}

1;
__END__

=head1 NAME

Encode::EUCJPMS - Microsoft Compatible Encodings for Japanese

=head1 SYNOPSIS

  use Encode::EUCJPMS;
  use Encode qw/encode decode/;
  $eucJP_ms = encode("eucJP-ms", $utf8);
  $utf8   = decode("eucJP-ms", $euc_jp);

=head1 ABSTRACT

This module implements Microsoft compatible encodings for Japanese.
Encodings supported are as follows.

  Canonical     Alias                                      Description
  --------------------------------------------------------------------
  eucJP-ms      qr/\beuc-?jp-?ms$/i                           eucJP-ms
                qr/\beuc-?jp-?win$/i
  cp51932                                       Windows Codepage 51932
  cp50220                                       Windows Codepage 50220
  cp50221                                       Windows Codepage 50221
  --------------------------------------------------------------------

=head1 DESCRIPTION

To find out how to use this module in detail, see L<Encode>.

=head1 INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

If you want to add eucJP-ms to Encode's demand-loading list
(so you don't have to "use Encode::EUCJPMS"), run

  enc2xs -C

to update Encode::ConfigLocal, a module that controls local settings.
After that, "use Encode;" is enough to load eucJP-ms on demand.

=head1 DEPENDENCIES

This module requires perl version 5.7.3 or later.

=head1 AUTHOR

NARUSE, Yui E<lt>naruse@airemix.comE<gt>

=head1 COPYRIGHT

Copyright (C) 2005-2006 NARUSE, Yui E<lt>naruse@airemix.comE<gt>

This program is free software; you can redistribute it and/or 
modify it under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=head1 SEE ALSO

L<Encode>, L<Encode::JP>

Problems and Solutions for Unicode and User/Vendor Defined Characters
L<http://www.opengroup.or.jp/jvc/cde/ucs-conv-e.html>

Windows Codepage 932
L<http://www.microsoft.com/globaldev/reference/dbcs/932.mspx>

=cut
